File: DSA/Stack/Code/InfixToPostfix.py
# Infix to Postfix in Python

def precedence(c):
    if c in ['+', '-']:
        return 1
    if c in ['*', '/']:
        return 2
    if c == '^':
        return 3
    return 0

def is_operator(c):
    return c in ['+', '-', '*', '/', '^']

def infix_to_postfix(infix):
    stack = []
    postfix = ""
    
    for c in infix:
        if c.isalnum():
            postfix += c
        elif c == '(':
            stack.append(c)
        elif c == ')':
            while stack and stack[-1] != '(':
                postfix += stack.pop()
            stack.pop()  # Remove '('
        elif is_operator(c):
            while stack and precedence(stack[-1]) >= precedence(c):
                postfix += stack.pop()
            stack.append(c)
    
    while stack:
        postfix += stack.pop()
    
    return postfix

# Test
infix = "a+b*c-d/e"

print("=== Infix to Postfix ===\n")
print(f"Infix Expression: {infix}")

postfix = infix_to_postfix(infix)

print(f"Postfix Expression: {postfix}\n")

print("Algorithm: Shunting-yard")
print("Complexity: O(n)")

File: DSA/Stack/Code/InfixToPostfix.cpp
#include <iostream>
#include <string>
#include <cctype>
#include <stack>
using namespace std;

// Infix to Postfix in C++
int precedence(char c) {
    if (c == '+' || c == '-') return 1;
    if (c == '*' || c == '/') return 2;
    if (c == '^') return 3;
    return 0;
}

bool isOperator(char c) {
    return (c == '+' || c == '-' || c == '*' || c == '/' || c == '^');
}

string infixToPostfix(string infix) {
    stack<char> s;
    string postfix = "";
    
    for (char c : infix) {
        if (isalnum(c)) {
            postfix += c;
        } else if (c == '(') {
            s.push(c);
        } else if (c == ')') {
            while (!s.empty() && s.top() != '(') {
                postfix += s.top();
                s.pop();
            }
            s.pop(); // Remove '('
        } else if (isOperator(c)) {
            while (!s.empty() && precedence(s.top()) >= precedence(c)) {
                postfix += s.top();
                s.pop();
            }
            s.push(c);
        }
    }
    
    while (!s.empty()) {
        postfix += s.top();
        s.pop();
    }
    
    return postfix;
}

int main() {
    string infix = "a+b*c-d/e";
    
    cout << "=== Infix to Postfix ===\n\n";
    cout << "Infix Expression: " << infix << "\n";
    
    string postfix = infixToPostfix(infix);
    
    cout << "Postfix Expression: " << postfix << "\n\n";
    
    cout << "Algorithm: Shunting-yard\n";
    cout << "Complexity: O(n)\n";
    
    return 0;
}
